sspl_test/coverage/coverage_setup.py

```diff
@@ -0,0 +1,172 @@
+from framework.base.sspl_constants import DATA_PATH
+
+file_path = "/opt/seagate/cortx/sspl/low-level"
+report_path = f"{DATA_PATH}coverage/sspl_xml_coverage_report.xml"
+
+PATCH_1 = """\
+from coverage import Coverage
+co = Coverage(
+        data_file='%scoverage/.sspl_coverage_report',
+        include="/opt/seagate/*",
+        omit=['*/.local/*', '*/usr/*'],
+        # config_file='%scoverage/.coveragerc',
+        )\
+"""%(DATA_PATH, DATA_PATH)
+
+PATCH_2 = """\
+        logger.info("Starting coverage report scope")
+        co.start()\
+"""
+
+PATCH_3 = """\
+def generate_cov_report(signal_number, frame):
+    logger.info('Ending Coverage Scope')
+    co.stop()
+    logger.info('coverage object stopped.')
+    co.save()
+    logger.info('coverage info saved.')
+    cov_per = co.xml_report(ignore_errors=True,
+        outfile='%scoverage/sspl_xml_coverage_report.xml')
+    logger.info(f'XML coverage report generated with coverage of {cov_per} percentage.')
+    ## Enable below code to inable HTML report generation
+ 	# html_cov_per = co.html_report(
+ 	#                     directory='%scoverage/sspl_html_coverage',
+ 	#                     ignore_errors=True,
+ 	#                 )
+ 	# logger.info(f'HTML coverage report geverated with coverage of {html_cov_per} percentage.')\
+"""%(DATA_PATH, DATA_PATH)
+
+PATCH_4 = """\
+    signal.signal(signal.SIGUSR1, generate_cov_report)\
+"""
+
+def coverage_setup():
+    """Installs pip3 coverage package. Injects different patches code from
+       coverage_code file to sspl_ll_d file, also creates target directory
+       for code coverage report and assigns permission to the directory.
+    """
+    print("Installing coverage.py")
+    _, err, return_code = SimpleProcess('python3 -m pip install coverage').run()
+    if return_code:
+        print(err)
+        return return_code
+
+    print("Creating required files for coverage..")
+    patch1_name = "initialize coverage obj for code coverage report generation"
+    patch2_name = "start the code coverage scope"
+    patch3_name = "stop coverage, save and generate code coverage report"
+    patch4_name = "signal handler for SIGUSR1 to generate code coverage report"
+
+    with open(f'{file_path}/sspl_ll_d', 'r') as sspl_ll_d:
+        sspl_ll_d_lines = sspl_ll_d.readlines()
+
+    for i, line in enumerate(sspl_ll_d_lines):
+        if patch1_name in line:
+            for j,l in enumerate(PATCH_1.split('\n')):
+                sspl_ll_d_lines.insert(i+j+1, l+'\n')
+        elif patch2_name in line:
+            for j,l in enumerate(PATCH_2.split('\n')):
+                sspl_ll_d_lines.insert(i+j+1, l+'\n')
+        elif patch3_name in line:
+            for j,l in enumerate(PATCH_3.split('\n')):
+                sspl_ll_d_lines.insert(i+j+1, l+'\n')
+        elif patch4_name in line:
+            for j,l in enumerate(PATCH_4.split('\n')):
+                sspl_ll_d_lines.insert(i+j+1, l+'\n')
+
+    shutil.move(f'{file_path}/sspl_ll_d',
```

Comment:
It should be 'copy' operation & not move here i believe.

Also, you are moving sspl_ll_d while its open already !? Not safe or wuld have unpredictable behavior !

Create sspl_ll_d.bak first, then open sspl_ll_d & update

Reply:
The file sspl_ll_d is not open.
on line 88,89 we are storing all the lines from the file in a list and then inserting lines in that list not the actual file.
then we are moving the sspl_ll_d file to .bak and then writing a new file sspl_ll_d file modified lines from the list.